src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java

```diff
@@ -83,15 +85,15 @@ public XContentType format() {
      * it's target filename of the pattern <tt>{prefix}{version}.st</tt>.
      *
      * @param state the state object to write
-     * @param prefix the state names prefix used to compose the file name.
      * @param version the version of the state
      * @param locations the locations where the state should be written to.
      * @throws IOException if an IOException occurs
      */
-    public final void write(final T state, final String prefix, final long version, final Path... locations) throws IOException {
+    public final void write(final T state, final long version, final Path... locations) throws IOException {
         Preconditions.checkArgument(locations != null, "Locations must not be null");
         Preconditions.checkArgument(locations.length > 0, "One or more locations required");
-        String fileName = prefix + version + STATE_FILE_EXTENSION;
+        final long maxStateId = Math.max(findMaxStateId(prefix, locations)+1, version);
```

Comment:
I don't feel comfortable with putting the version in the mix. I think I see why - the file name will be the version most of the time, but it is not guaranteed which makes me wonder if it's worth it. Maybe we should have prefix.version.id.state  where id is always incremented?